File: leetcode/JavaScript/No139.word-break.js
/**
 * Difficulty:
 * Medium
 *
 * Desc:
 * Given a non-empty string s and a dictionary wordDict containing a list of non-empty words,
 * determine if s can be segmented into a space-separated sequence of one or more dictionary words.
 * You may assume the dictionary does not contain duplicate words.
 *
 * Example:
 * Given
 * s = "leetcode",
 * dict = ["leet", "code"].
 * Return true because "leetcode" can be segmented as "leet code".
 *
 * 给定一个非空字符串 s 和一个包含非空单词列表的字典 wordDict，判定 s 是否可以被空格拆分为一个或多个在字典中出现的单词。
 * 说明：
 * 1. 拆分时可以重复使用字典中的单词。
 * 2. 你可以假设字典中没有重复的单词
 */

/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 *
 * DFS
 */
var wordBreak = function(s, wordDict) {
  const set = new Set(wordDict)
  const tmp = {}

  const check = (index) => {
    if (tmp[index] !== undefined) return tmp[index]
    if (index >= s.length) return true

    for (let i = index; i < s.length; i += 1) {
      if (set.has(s.slice(index, i + 1))) {
        tmp[i + 1] = check(i + 1)
        if (tmp[i + 1]) {
          return true
        }
      }
    }
    tmp[index] = false
    return false
  }

  return check(0)
}

/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 *
 * Dynamic Programming
 */
const wordBreak_2 = (s, wordDict) => {
  const set = new Set(wordDict)
  const tmp = [true]

  for (let i = 1; i <= s.length; i += 1) {
    for (let j = 0; j < i; j += 1) {
      if (tmp[j] && set.has(s.slice(j, i))) {
        tmp[i] = true
        break
      }
    }
  }

  return tmp[s.length] === true
}

File: leetcode/C#/No139.word-break.cs
/*
 * Difficulty:
 * Medium
 *
 * Desc:
 * Given a non-empty string s and a dictionary wordDict containing a list of non-empty words,
 * determine if s can be segmented into a space-separated sequence of one or more dictionary words.
 *
 * Note:
 * The same word in the dictionary may be reused multiple times in the segmentation.
 * You may assume the dictionary does not contain duplicate words.
 *
 * Example 1:
 * Input: s = "leetcode", wordDict = ["leet", "code"]
 * Output: true
 * Explanation: Return true because "leetcode" can be segmented as "leet code".
 *
 * Example 2:
 * Input: s = "applepenapple", wordDict = ["apple", "pen"]
 * Output: true
 * Explanation:
 * Return true because "applepenapple" can be segmented as "apple pen apple".
 * Note that you are allowed to reuse a dictionary word.
 *
 * Example 3:
 * Input: s = "catsandog", wordDict = ["cats", "dog", "sand", "and", "cat"]
 * Output: false
*/

public class Solution {
    public bool WordBreak(string s, IList<string> wordDict) {
        bool[] dp = new bool[s.Length + 1];
        dp[0] = true;

        HashSet<string> set = new HashSet<string>(wordDict);
        for (int i = 1; i <= s.Length; i += 1) {
            for (int j = i - 1; j >= 0; j -= 1) {
                dp[i] = set.Contains(s[j..i]) && dp[j];
                if (dp[i]) break;
            }
        }
        return dp[^1];
    }
}